### PR TITLE
Fix #10444: Client Window Factory

### LINKED ISSUE #10444: JSF 2.2 removal: use client-window-factory
in the past we used lifecyclefactory to have JSF 2.2 compatiblity, now we can use client-window-factory directly
it seems it doesnt work at least in Mojarra 2.3, Myfaces 2.3 works fine

@melloware are you interessted to check that?

simple add a PrimeClientWindowFactory and return new PrimeClientWindow + register it in faces-config

### PR DESCRIPTION
**Pull request by melloware**
Fix #10444: Client Window Factory

### CODE REVIEW

**Comment by melloware** [comment]
@tandraschko 2 things.

1. Mojarra requires the web.xml context param `javax.faces.CLIENT_WINDOW_MODE` or it won't activate.
2. I am getting a stack trace that gets ignored...

```
java.lang.IllegalStateException: Committed
    at org.eclipse.jetty.server.HttpChannel.resetBuffer (HttpChannel.java:1093)
    at org.eclipse.jetty.server.HttpOutput.resetBuffer (HttpOutput.java:1467)
    at org.eclipse.jetty.server.Response.resetBuffer (Response.java:1219)
    at org.eclipse.jetty.server.Response.sendRedirect (Response.java:571)
    at org.eclipse.jetty.server.Response.sendRedirect (Response.java:505)
    at org.eclipse.jetty.server.Response.sendRedirect (Response.java:580)
    at com.sun.faces.context.ExternalContextImpl.redirect (ExternalContextImpl.java:725)
    at javax.faces.context.ExternalContextWrapper.redirect (ExternalContextWrapper.java:597)
    at org.primefaces.clientwindow.PrimeClientWindow.decode (PrimeClientWindow.java:82)
    at com.sun.faces.lifecycle.LifecycleImpl.attachWindow (LifecycleImpl.java:131)
    at javax.faces.webapp.FacesServlet.executeLifecyle (FacesServlet.java:706)
    at javax.faces.webapp.FacesServlet.service (FacesServlet.java:451)
```

File: primefaces/src/main/java/org/primefaces/clientwindow/PrimeClientWindowFactory.java [hunk 1]
```diff
@@ -25,31 +25,28 @@
...
+    public ClientWindow getClientWindow(FacesContext context) {
+        ClientWindow clientWindow = context.getExternalContext().getClientWindow();
         if (clientWindow == null) {
             clientWindow = new PrimeClientWindow();
         }
-
         try {
-            clientWindow.decode(facesContext);
-            facesContext.getExternalContext().setClientWindow(clientWindow);
+            clientWindow.decode(context);
```
Comment (melloware) [inline primefaces/src/main/java/org/primefaces/clientwindow/PrimeClientWindowFactory.java]: I think this logic needs to be examined whether it is still needed?

File: primefaces/src/main/java/org/primefaces/clientwindow/PrimeClientWindow.java [hunk 2]
```diff
@@ -84,8 +85,6 @@ public void decode(FacesContext context) {
                 catch (IOException e) {
                     throw new FacesException("Could not send initial redirect!", e);
                 }
-
-                context.responseComplete();
```
Comment (melloware) [inline primefaces/src/main/java/org/primefaces/clientwindow/PrimeClientWindow.java]: Had to remove this as its no longer part of LifeCycle and it kept reporting "response is closed".

**Comment by melloware** [comment]
I don't think we are allowed to redirect anymore either because its not a lifecycle?

**Comment by tandraschko** [comment]
Actually we must NOT change the ClientWindow itself, only the factory part
Before we attached it via a own lifecycle because 2.2
Now we can simple use the real factory actually

**Comment by tandraschko** [comment]
It also worked fine on MyFaces 2.3

**Comment by melloware** [comment]
OK updated my PR now all integration tests are passing and I submitted a Mojarra ticket about needing the `web.xml` param to enable our custom factory.

**Comment by tandraschko** [comment]
should we mention that in the docs, that only Mojarra requires the context-param?

**Comment by melloware** [comment]
Note added about Mojarra and link provided.

**Comment by melloware** [comment]
@tandraschko I think I am seeing one strange this with this.  Now with MyFaces I see it get windowID 1234 then immediately refresh and get 4567 for example. Like on first page load its double loading.  Do you think it is because of the web.xml `javax.faces.CLIENT_WINDOW_MODE` it seems like Myfaces default handling is happening and then our custom on kicks in?

**Comment by tandraschko** [comment]
can you try to remove javax.faces.CLIENT_WINDOW_MODE and see what happens?

**Comment by melloware** [comment]
Yes I will test and report back.

**Comment by melloware** [comment]
Nevermind, MyFaces is doing the right thing its Mojarra that does the double redirect but it all related to that issue I already reported. Sorry!

**Comment by tandraschko** [comment]
but that means that we can't remove the old LifecycleFactory :/

I think we should leave it and provide the alternative ClientWIndowFactory

then we can mention in the docs: Use ClientWindowFactory but in mojarra older then version x, please use the old lifecyclefactory

**Comment by melloware** [comment]
OK I have put the LifeCycle back in as well and updated the docs to mention use direct Factory for MyFaces and LifeCycleFactory for Mojarra.

File: primefaces-integration-tests-jakarta/src/main/webapp/WEB-INF/web.xml [hunk 3]
```diff
@@ -18,6 +18,10 @@
         <param-name>jakarta.faces.STATE_SAVING_METHOD</param-name>
         <param-value>${jakarta.faces.STATE_SAVING_METHOD}</param-value>
     </context-param>
+    <context-param>
+        <param-name>jakarta.faces.CLIENT_WINDOW_MODE</param-name>
```
Comment (tandraschko) [inline primefaces-integration-tests-jakarta/src/main/webapp/WEB-INF/web.xml]: shouldnt we remove this?

File: primefaces-integration-tests/src/main/webapp/WEB-INF/web.xml [hunk 4]
```diff
@@ -18,6 +18,10 @@
         <param-name>javax.faces.STATE_SAVING_METHOD</param-name>
         <param-value>${javax.faces.STATE_SAVING_METHOD}</param-value>
     </context-param>
+    <context-param>
+        <param-name>javax.faces.CLIENT_WINDOW_MODE</param-name>
```
Comment (tandraschko) [inline primefaces-integration-tests/src/main/webapp/WEB-INF/web.xml]: shouldnt we remove this?

File: primefaces-showcase/src/main/webapp/WEB-INF/web.xml [hunk 5]
```diff
@@ -37,6 +37,11 @@
         <param-name>javax.faces.INTERPRET_EMPTY_STRING_SUBMITTED_VALUES_AS_NULL</param-name>
         <param-value>true</param-value>
     </context-param>
+    <!-- enable client window mode -->
+    <context-param>
+        <param-name>javax.faces.CLIENT_WINDOW_MODE</param-name>
```
Comment (tandraschko) [inline primefaces-showcase/src/main/webapp/WEB-INF/web.xml]: shouldnt we remove this?

Comment (melloware) [reply primefaces-showcase/src/main/webapp/WEB-INF/web.xml]: We can remove it from showcase but it means you can only ever run the showcase with MyFaces which is OK with me?

File: primefaces-integration-tests/src/main/webapp/WEB-INF/web.xml (hunk 4 above)
Comment (melloware) [reply primefaces-integration-tests/src/main/webapp/WEB-INF/web.xml]: Until Mojarra fixes it the integration tests don't work without this unless I put the LifeCycleMapping back in the faces-config.xml of the Integration tests?

File: primefaces-showcase/src/main/webapp/WEB-INF/web.xml (hunk 5 above)
Comment (tandraschko) [reply primefaces-showcase/src/main/webapp/WEB-INF/web.xml]: AFAIR you said this param even breaks Mojarra?
So i would just use the LifecycleFactory in our showcase/IT's but suggest the new one for MF users

Comment (melloware) [reply primefaces-showcase/src/main/webapp/WEB-INF/web.xml]: OK i will revert these to use LifeCycle and then it will work for both again.

**Comment by tandraschko** [comment]
+1

**Comment by melloware** [comment]
I messed up my branch so closing and re-opening.

**Pull request closed without merging**